File: solutions/python3/0061.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next or k == 0:
            return head

        len = 0
        curr = head
        while curr:
            len += 1
            curr = curr.next

        k %= len
        if k == 0:
            return head

        slow = head
        fast = head

        for _ in range(k):
            fast = fast.next

        while fast and fast.next:
            slow = slow.next
            fast = fast.next

        ans = slow.next
        slow.next = None
        fast.next = head

        return ans

File: solutions/cpp/0061.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) return head;

        int len = 0;
        for (auto curr = head; curr; curr = curr->next) len++;

        k %= len;
        if (k == 0) return head;

        auto slow = head;
        auto fast = head;

        while (k--) fast = fast->next;

        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next;
        }

        auto ans = slow->next;
        slow->next = NULL;
        fast->next = head;

        return ans;
    }
};